lib/puppet/provider/exec.rb

```diff
@@ -26,6 +26,10 @@ def environment
         warning _("Overriding environment setting '%{var}' with '%{value}'") % { var: var, value: value }
       end
 
+      if value.nil? || value.empty?
+        warning _("Empty environment setting '%{var}'") % {var: var}
```

Comment:
I think it might be more useful to warn about these unset variables in a way that folks can turn off. I need to get confirmation that this is the best way, but something like this:
```suggestion
        warn_msg = _("Empty environment setting '%{var}'") % {var: var}
        Puppet.warn_once('undefined_variables', 'empty_env_var', warn_msg)
```

This way, folks can turn off the warning and declutter their output
```
$ be ./bin/puppet apply -e "exec{'foobar': command => '/usr/bin/true', environment => 'FOO='}"
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Warning: Empty environment setting 'FOO'
   (file & line not available)
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.02 seconds
```
```
$ be ./bin/puppet apply -e "exec{'foobar': command => '/usr/bin/true', environment => 'FOO='}" --disable_warnings undefined_variables
Notice: Compiled catalog for bubba.local in environment production in 0.07 seconds
Notice: /Stage[main]/Main/Exec[foobar]/returns: executed successfully
Notice: Applied catalog in 0.03 seconds
```

Reply:
Hi @melissa. Thanks for your review. I agree with your suggestion and would be happy to update my PR.

Please let me know once you can confirm it's okay.

Reply:
Wonderful!

We might want to add the file and line, if they're available to the provider. I didn't look into this at all, but I'm happy to help if you'd like.

The `warn_once` method is https://github.com/puppetlabs/puppet/blob/0f57fdf115604c6fd3924aa064bd8f6d964c797e/lib/puppet/util/logging.rb#L168

We probably should make `key` unique for each different environment var, rather than have it the same for any missing env var. According to the warn_once docs, `@param key [String] Key used to make this warning unique`. So you might want to change `'empty_env_var'` in my example to something like `"empty_env_var_#{var}"`. I haven't played with that to make sure it will work, though I expect it will.

Feel free to ping me if you want any guidance on these changes. I'm happy to help out!